modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLTexture.java

```diff
@@ -0,0 +1,253 @@
+        this.texPtr = sharedTex.texPtr;
+    }
+
+    @Override
+    protected Texture createSharedTexture(WrapMode newMode) {
+        return new MTLTexture(this, newMode);
+    }
+
+    private void updateTextureInt(Buffer buffer, PixelFormat format,
+                                int dstx, int dsty,
+                                int srcx, int srcy,
+                                int srcw, int srch,
+                                int srcscan) {
+        if (format == PixelFormat.INT_ARGB_PRE) {
+            IntBuffer buf = (IntBuffer) buffer;
+            int[] arr = buf.hasArray() ? buf.array() : null;
+            nUpdateInt(getNativeHandle(), buf, arr,
+                       dstx, dsty, srcx, srcy, srcw, srch, srcscan);
+        } else {
+            throw new IllegalArgumentException("Unsupported INT PixelFormat: " + format);
+        }
+    }
+
+    private void updateTextureFloat(Buffer buffer, PixelFormat format,
+                                int dstx, int dsty,
+                                int srcx, int srcy,
+                                int srcw, int srch,
+                                int srcscan) {
+        if (format == PixelFormat.FLOAT_XYZW) {
+            FloatBuffer buf = (FloatBuffer) buffer;
+            float[] arr = buf.hasArray() ? buf.array() : null;
+            nUpdateFloat(getNativeHandle(), buf, arr,
+                         dstx, dsty, srcx, srcy, srcw, srch, srcscan);
+        } else {
+            throw new IllegalArgumentException("Unsupported FLOAT PixelFormat: " + format);
+        }
+    }
+
+    private void updateTextureByte(Buffer buffer, PixelFormat format,
+                                int dstx, int dsty,
+                                int srcx, int srcy,
+                                int srcw, int srch,
+                                int srcscan) {
+        ByteBuffer buf = (ByteBuffer) buffer;
+        buf.rewind();
+        byte[] arr = buf.hasArray() ? buf.array() : null;
+
+        switch (format) {
+            case PixelFormat.BYTE_BGRA_PRE,
+                 PixelFormat.BYTE_ALPHA ->
+                nUpdate(getNativeHandle(), buf, arr,
+                        dstx, dsty, srcx, srcy, srcw, srch, srcscan);
+
+            case PixelFormat.BYTE_RGB -> {
+                // Convert 24-bit RGB to 32-bit BGRA
+                // Metal does not support 24-bit format
+                // hence `arr` data needs to be converted to BGRA format
+                byte[] arr32Bit = new byte[srcw * srch * 4];
+                int dstIndex = 0;
+                int index = 0;
+
+                int rowStride = srcw * 3;
+                int totalBytes = srch * rowStride;
+
+                for (int rowIndex = 0; rowIndex < totalBytes; rowIndex += rowStride) {
+                    for (int colIndex = 0; colIndex < rowStride; colIndex += 3) {
+                        index = rowIndex + colIndex;
+                        arr32Bit[dstIndex++] = arr[index + 2];
```

Comment:
`arr` might be `null` here if `buf.hasArray() == false`, resulting in an NPE.

Reply:
The buffer is sometimes direct, so the fallback case of creating a new byte array and copying it from the buffer should only be done where it is needed. See my inline comment.